scipy/sparse/_coo.py

```diff
@@ -528,8 +529,8 @@ def _add_dense(self, other):
         dtype = upcast_char(self.dtype.char, other.dtype.char)
         result = np.array(other, dtype=dtype, copy=True)
         fortran = int(result.flags.f_contiguous)
-        M = self.shape[0]
-        N = self.shape[1] if self.ndim > 1 else 1
+        M = self.shape[0] if self.ndim > 1 else 1
+        N = self.shape[-1]
```

Comment:
Alternately:

```
*_, M, N = (1, 1) + self.shape
```

Maybe we should have private helper method for getting `(M, N)`, so we don't have to remember the trick every time.

Reply:
I also thought about have a helper function, but there do seem to be multiple desires:
- finding out if this object is 2d or 1d
- reporting separately the length of each dimension
- reporting the 2d-version of the length of each dimension (for calling the sparsetools functions)

Perhaps we should have an attribute/property `self._shape2d`.  This would/will be needed with nd representations too -- at least the ones that represent the nd-sparse as a 2d-sparse like GCRS/GCRC or others. And it would do much of what these idioms here would do.

So, we could use have 3 properties/attributes `ndim` to tell 1d from 2d, `shape` to give the actual shape, and `shape2d` to give the 2d representation needed for  sparsetools functions.